File: LeetCodeRecord/MethodFirst/79-Exist.java
package MethodFirst;

class Exist {
    private boolean[][] marked;
    private int rows;
    private int columns;

    public boolean exist(char[][] board, String word) {
        this.rows = board.length;
        this.columns = board[0].length;

        this.marked = new boolean[this.rows][this.columns];

        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.columns; j++) {
                if (dfs(i, j, board, 0, word)) {
                    return true;
                }
            }
        }

        return false;
    }

    public boolean dfs(int row, int column, char[][] board, int wordIndex, String word) {
        if (row < 0 || row >= this.rows) {
            return false;
        }

        if (column < 0 || column >= this.columns) {
            return false;
        }

        if (this.marked[row][column] == true) {
            return false;
        }

        if (wordIndex == word.length() - 1) {
            return board[row][column] == word.charAt(wordIndex);
        }

        if (word.charAt(wordIndex) == board[row][column]) {
            this.marked[row][column] = true;
            
            if (dfs(row - 1, column, board, wordIndex + 1, word)) {
                return true;
            }

            if (dfs(row, column + 1, board, wordIndex + 1, word)) {
                return true;
            }

            if (dfs(row + 1, column, board, wordIndex + 1, word)) {
                return true;
            }

            if (dfs(row, column - 1, board, wordIndex + 1, word)) {
                return true;
            }
        }

        this.marked[row][column] =  false;

        return false;
    }
}

File: LeetCodeRecord/ACFirst-Golang/79-Exist.go
package main

var visited [][]bool
var rows int
var columns int

func exist(board [][]byte, word string) bool {
	rows = len(board)
	if rows == 0 {
		return false
	}
	columns = len(board[0])

	visited = make([][]bool, rows)

	for i := 0; i < rows; i++ {
		visited[i] = make([]bool, columns)
	}

	for i := 0; i < rows; i++ {
		for j := 0; j < columns; j++ {
			if dfs(board, i, j, word, 0) {
				return true
			}
		}
	}

	return false
}

func dfs(board [][]byte, row, column int, word string, index int) bool {
	if row < 0 || row >= rows {
		return false
	}

	if column < 0 || column >= columns {
		return false
	}

	if visited[row][column] {
		return false
	}

	if board[row][column] != word[index] {
		return false
	}

	if index == len(word)-1 {
		return true
	}

	visited[row][column] = true

	if dfs(board, row+1, column, word, index+1) {
		return true
	}

	if dfs(board, row-1, column, word, index+1) {
		return true
	}

	if dfs(board, row, column+1, word, index+1) {
		return true
	}

	if dfs(board, row, column-1, word, index+1) {
		return true
	}

	visited[row][column] = false

	return false
}
